lib/DateTime/Duration.pm

```diff
@@ -64,7 +64,7 @@ my @all_units = qw( months days minutes seconds nanoseconds );
 
     sub new {
         my $class = shift;
-        my %p     = $check->(@_);
+        my %p     = $check->( ref($_[0]) eq $class ? %{ $_[0] } : @_);
```

Comment:
Fair enough; I was aiming for the minimal changes required to keep this feature working the way it always used to. Will look at a tidier patch.